test/model/recent_senders_test.dart

```diff
@@ -0,0 +1,229 @@
+        }
+      }
+
+      test('batch goes before the existing messages', () {
+        final existingMessages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 300),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 400),
+        ];
+        setupModel(existingMessages);
+
+        final messages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 100),
+          eg.streamMessage(stream: stream1, topic: 'b', sender: user10, id: 200),
+        ];
+        model.handleMessages(messages);
+
+        checkMatchesMessages(model,
+          [...existingMessages, ...messages]..sort((m1, m2) => m1.id.compareTo(m2.id)));
+      });
+
+      test('batch goes after the existing messages', () {
+        final existingMessages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 300),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 400),
+        ];
+        setupModel(existingMessages);
+
+        final messages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 500),
+          eg.streamMessage(stream: stream1, topic: 'b', sender: user10, id: 600),
+        ];
+        model.handleMessages(messages);
+
+        checkMatchesMessages(model,
+          [...existingMessages, ...messages]..sort((m1, m2) => m1.id.compareTo(m2.id)));
+      });
+
+      test('batch is interspersed among the existing messages', () {
+        final existingMessages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 200),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 400),
+        ];
+        setupModel(existingMessages);
+
+        final messages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 100),
+          eg.streamMessage(stream: stream1, topic: 'b', sender: user10, id: 300),
+          eg.streamMessage(stream: stream1, topic: 'b', sender: user10, id: 500),
+        ];
+        model.handleMessages(messages);
+
+        checkMatchesMessages(model,
+          [...existingMessages, ...messages]..sort((m1, m2) => m1.id.compareTo(m2.id)));
+      });
+
+      test('batch contains some of already-existing messages', () {
+        final existingMessages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 200),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 300),
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 400),
+        ];
+        setupModel(existingMessages);
+
+        final messages = [
+          eg.streamMessage(stream: stream1, topic: 'a', sender: user10, id: 100),
+          eg.streamMessage(stream: stream1, topic: 'b', sender: user10, id: 200),
+          eg.streamMessage(stream: stream1, topic: 'b', sender: user10, id: 400),
+          eg.streamMessage(stream: stream1, topic: 'b', sender: user10, id: 500),
```

Comment:
Similarly here. Though this one has another problem, which is that the data is unrealistic: some of these are the same message IDs again but in a different place.

(That isn't actually completely impossible, given the fetch-vs-event race: the messages could have moved. But if we were going to write a test that exercised that, it'd be a test that was focused on that scenario. And before #150 it won't make sense to try to test that anyway.)

Reply:
As before, these messages were also meant to be in `topic: 'a'`. 🙂 Thanks for catching.